jobs/jobs.go

```diff
@@ -1,35 +1,78 @@
 package jobs
 
-import "sync"
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
 
 // Job represents a (long-running) process that can be spawned on a separate go-routine.
 // When encountering an error, a Job should send that error to the given channel and continue to run.
-type Job func(chan<- error)
+type Job func(ctx context.Context) error
 
 // JobManager manages multiple concurrent jobs and handles their errors. Can wait for all jobs and error handling to finish.
 type JobManager struct {
-	wgJobs  *sync.WaitGroup
-	wgErr   *sync.WaitGroup
-	errChan chan error
+	wgJobs      *sync.WaitGroup
+	done        chan struct{}
+	errChan     chan error
+	ctx         context.Context
+	once        *sync.Once
+	jobCapacity *capacityMgr
+}
+
+type capacityMgr struct {
+	semaphore *semaphore.Weighted
+}
+
+func (mgr *capacityMgr) Acquire(ctx context.Context, n int64) error {
+	if mgr.semaphore != nil {
+		return mgr.semaphore.Acquire(ctx, n)
+	}
+	return nil
+}
+
+func (mgr *capacityMgr) Release(n int64) {
+	if mgr.semaphore != nil {
+		mgr.semaphore.Release(n)
+	}
+}
+
+// MgrOptions modify the behaviour of the JobManager
+type MgrOptions func(*JobManager) *JobManager
+
+// WithMaxCapacity defines how many jobs will be run in parallel
+func WithMaxCapacity(cap int64) MgrOptions {
+	return func(mgr *JobManager) *JobManager {
+		mgr.jobCapacity = &capacityMgr{semaphore.NewWeighted(cap)}
+		return mgr
+	}
+}
+
+// WithErrorCacheCapacity defines the size of the error cache. If the cache is full new errors from jobs will be ignored. Default is 1000
+func WithErrorCacheCapacity(cap int64) MgrOptions {
+	return func(mgr *JobManager) *JobManager {
+		mgr.errChan = make(chan error, cap)
+		return mgr
+	}
 }
 
 // NewMgr returns a new JobManager
-func NewMgr(errHandler func(err error)) *JobManager {
-	wgErr := &sync.WaitGroup{}
-	wgErr.Add(1)
-	errChan := make(chan error, 1000)
-	go func() {
-		defer wgErr.Done()
-		for err := range errChan {
-			errHandler(err)
-		}
-	}()
+func NewMgr(ctx context.Context, opts ...MgrOptions) *JobManager {
+	mgr := &JobManager{
+		ctx:         ctx,
+		jobCapacity: &capacityMgr{},
+		done:        make(chan struct{}),
+		once:        &sync.Once{},
+		errChan:     make(chan error, 1000),
```

Comment:
shouldn't this be initialized to basically whatever jobCapacity is?

Reply:
job capacity can be infinite, so I'd like to keep them separate